In [1]:
import pandas as pd 
import numpy as np

In [2]:
df = pd.read_csv('https://s3.us.cloud-object-storage.appdomain.cloud/cf-courses-data/CognitiveClass/DP0701EN/version-2/Data-Collisions.csv')

/opt/conda/envs/Python36/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (33) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
import warnings
warnings.filterwarnings('ignore')

# Data Preprocessing 

In [4]:
df.head(5)

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  ... ROADCOND                LIGHTCOND  \
0  Matched  Intersection  37475.0  ...      Wet                 Daylight   
1  Matched         Block      NaN  ...      Wet  Dark - Street Lights On   
2  Matched         Block      NaN  ...      Dry                 Daylight   
3  Matched         Block      NaN  ...      Dry                 Daylight   
4  Matched  Intersection  34387.0  ...      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Entering at angle           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 38 columns]

In [5]:
df.dtypes

SEVERITYCODE        int64
X                 float64
Y                 float64
OBJECTID            int64
INCKEY              int64
COLDETKEY           int64
REPORTNO           object
STATUS             object
ADDRTYPE           object
INTKEY            float64
LOCATION           object
EXCEPTRSNCODE      object
EXCEPTRSNDESC      object
SEVERITYCODE.1      int64
SEVERITYDESC       object
COLLISIONTYPE      object
PERSONCOUNT         int64
PEDCOUNT            int64
PEDCYLCOUNT         int64
VEHCOUNT            int64
INCDATE            object
INCDTTM            object
JUNCTIONTYPE       object
SDOT_COLCODE        int64
SDOT_COLDESC       object
INATTENTIONIND     object
UNDERINFL          object
WEATHER            object
ROADCOND           object
LIGHTCOND          object
PEDROWNOTGRNT      object
SDOTCOLNUM        float64
SPEEDING           object
ST_COLCODE         object
ST_COLDESC         object
SEGLANEKEY          int64
CROSSWALKKEY        int64
HITPARKEDCAR       object
dtype: objec

In [6]:
df["SEVERITYCODE"].value_counts()

1    136485
2     58188
Name: SEVERITYCODE, dtype: int64

In [7]:
df["ROADCOND"].value_counts()

Dry               124510
Wet                47474
Unknown            15078
Ice                 1209
Snow/Slush          1004
Other                132
Standing Water       115
Sand/Mud/Dirt         75
Oil                   64
Name: ROADCOND, dtype: int64

In [8]:
df["LIGHTCOND"].value_counts()

Daylight                    116137
Dark - Street Lights On      48507
Unknown                      13473
Dusk                          5902
Dawn                          2502
Dark - No Street Lights       1537
Dark - Street Lights Off      1199
Other                          235
Dark - Unknown Lighting         11
Name: LIGHTCOND, dtype: int64

In [9]:
df["WEATHER"].value_counts()

Clear                       111135
Raining                      33145
Overcast                     27714
Unknown                      15091
Snowing                        907
Other                          832
Fog/Smog/Smoke                 569
Sleet/Hail/Freezing Rain       113
Blowing Sand/Dirt               56
Severe Crosswind                25
Partly Cloudy                    5
Name: WEATHER, dtype: int64

In [10]:
df2 = df[["SEVERITYCODE", "ROADCOND", "LIGHTCOND", "WEATHER"]]
df2.head(5)

SEVERITYCODE ROADCOND                LIGHTCOND   WEATHER
0             2      Wet                 Daylight  Overcast
1             1      Wet  Dark - Street Lights On   Raining
2             1      Dry                 Daylight  Overcast
3             1      Dry                 Daylight     Clear
4             2      Wet                 Daylight   Raining

In [11]:
df2.replace("Unknown", np.nan, inplace=True)
df2.head(20)

SEVERITYCODE ROADCOND                LIGHTCOND   WEATHER
0              2      Wet                 Daylight  Overcast
1              1      Wet  Dark - Street Lights On   Raining
2              1      Dry                 Daylight  Overcast
3              1      Dry                 Daylight     Clear
4              2      Wet                 Daylight   Raining
5              1      Dry                 Daylight     Clear
6              1      Wet                 Daylight   Raining
7              2      Dry                 Daylight     Clear
8              1      Dry                 Daylight     Clear
9              2      Dry                 Daylight     Clear
10             1      Dry                 Daylight  Overcast
11             1      Dry                 Daylight     Clear
12             1      Wet  Dark - Street Lights On   Raining
13             1      Wet  Dark - No Street Lights   Raining
14             2      Dry  Dark - Street Lights On     Clear
15             1      NaN                      NaN       NaN
16             2      Dry                 Daylight  Overcast
17             1      Dry                 Daylight  Overcast
18             2      Dry                 Daylight     Clear
19             1      Dry                      NaN       NaN

In [12]:
df2.dropna(subset=["LIGHTCOND", "ROADCOND", "WEATHER"], axis=0, inplace=True)

df2.reset_index(drop=True, inplace=True)

df2.head(20)

SEVERITYCODE ROADCOND                LIGHTCOND   WEATHER
0              2      Wet                 Daylight  Overcast
1              1      Wet  Dark - Street Lights On   Raining
2              1      Dry                 Daylight  Overcast
3              1      Dry                 Daylight     Clear
4              2      Wet                 Daylight   Raining
5              1      Dry                 Daylight     Clear
6              1      Wet                 Daylight   Raining
7              2      Dry                 Daylight     Clear
8              1      Dry                 Daylight     Clear
9              2      Dry                 Daylight     Clear
10             1      Dry                 Daylight  Overcast
11             1      Dry                 Daylight     Clear
12             1      Wet  Dark - Street Lights On   Raining
13             1      Wet  Dark - No Street Lights   Raining
14             2      Dry  Dark - Street Lights On     Clear
15             2      Dry                 Daylight  Overcast
16             1      Dry                 Daylight  Overcast
17             2      Dry                 Daylight     Clear
18             2      Dry  Dark - Street Lights On     Clear
19             2      Dry                 Daylight     Clear

In [13]:
df2["SEVERITYCODE"].value_counts()

1    114659
2     55851
Name: SEVERITYCODE, dtype: int64

In [14]:
df2["WEATHER"] = df2["WEATHER"].astype('category')
df2["ROADCOND"] = df2["ROADCOND"].astype('category')
df2["LIGHTCOND"] = df2["LIGHTCOND"].astype('category')

In [15]:
from sklearn.utils import resample

In [16]:
df2_maj = df2[df2.SEVERITYCODE==1]
df2_min = df2[df2.SEVERITYCODE==2]

df2_maj_resample = resample(df2_maj, replace=False, n_samples=55851, random_state=123)

df3 = pd.concat([df2_maj_resample, df2_min])
df3.SEVERITYCODE.value_counts()

2    55851
1    55851
Name: SEVERITYCODE, dtype: int64

In [17]:
df3.reset_index(drop=True, inplace=True)
df3.head(10)

SEVERITYCODE ROADCOND                LIGHTCOND WEATHER
0             1      Dry                 Daylight   Clear
1             1      Dry  Dark - Street Lights On   Clear
2             1      Dry                 Daylight   Clear
3             1      Wet                 Daylight   Clear
4             1      Dry  Dark - Street Lights On   Clear
5             1      Dry                 Daylight   Clear
6             1      Dry  Dark - Street Lights On   Clear
7             1      Dry                 Daylight   Clear
8             1      Dry                 Daylight   Clear
9             1      Dry                 Daylight   Clear

In [18]:
df3.tail(10)

SEVERITYCODE ROADCOND                LIGHTCOND   WEATHER
111692             2      Dry                 Daylight     Clear
111693             2      Wet  Dark - Street Lights On  Overcast
111694             2      Dry                 Daylight     Clear
111695             2      Wet                    Other   Raining
111696             2      Dry                 Daylight     Clear
111697             2      Wet                 Daylight   Raining
111698             2      Wet                 Daylight     Clear
111699             2      Dry                 Daylight     Clear
111700             2      Dry                 Daylight     Clear
111701             2      Dry                     Dusk     Clear

In [19]:
df3["WEATHER_NUM"] = df3["WEATHER"].cat.codes
df3["ROADCOND_NUM"] = df3["ROADCOND"].cat.codes
df3["LIGHTCOND_NUM"] = df3["LIGHTCOND"].cat.codes

In [20]:
df3.head(10)

SEVERITYCODE ROADCOND                LIGHTCOND WEATHER  WEATHER_NUM  \
0             1      Dry                 Daylight   Clear            1   
1             1      Dry  Dark - Street Lights On   Clear            1   
2             1      Dry                 Daylight   Clear            1   
3             1      Wet                 Daylight   Clear            1   
4             1      Dry  Dark - Street Lights On   Clear            1   
5             1      Dry                 Daylight   Clear            1   
6             1      Dry  Dark - Street Lights On   Clear            1   
7             1      Dry                 Daylight   Clear            1   
8             1      Dry                 Daylight   Clear            1   
9             1      Dry                 Daylight   Clear            1   

   ROADCOND_NUM  LIGHTCOND_NUM  
0             0              5  
1             0              2  
2             0              5  
3             7              5  
4             0              2  
5             0              5  
6             0              2  
7             0              5  
8             0              5  
9             0              5

In [21]:
X = np.asarray(df3[['WEATHER_NUM', 'ROADCOND_NUM', 'LIGHTCOND_NUM']])
X[0:5]

array([[1, 0, 5],
       [1, 0, 2],
       [1, 0, 5],
       [1, 7, 5],
       [1, 0, 2]], dtype=int8)

In [22]:
y = np.asarray(df3['SEVERITYCODE'])
y [0:5]

array([1, 1, 1, 1, 1])

In [23]:
from sklearn import preprocessing
X = preprocessing.StandardScaler().fit(X).transform(X)
X[0:5]

array([[-0.71024412, -0.625086  ,  0.59417262],
       [-0.71024412, -0.625086  , -1.50252518],
       [-0.71024412, -0.625086  ,  0.59417262],
       [-0.71024412,  1.61670159,  0.59417262],
       [-0.71024412, -0.625086  , -1.50252518]])

# Modeling and Evaluation 

In [24]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss

## Train and Test 

In [25]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=4)
print ('Train set rows:', X_train.shape[0])
print ('Test set rows:', X_test.shape[0])

Train set rows: 78191
Test set rows: 33511


## KNN

In [26]:
 from sklearn.neighbors import KNeighborsClassifier
k = 5
knn = KNeighborsClassifier(n_neighbors = k).fit(X_train,y_train)

knn_y_pred = knn.predict(X_test)
knn_y_pred[0:5]

array([2, 2, 2, 2, 1])

In [27]:
Ks=20
mean_acc=np.zeros((Ks-1))
std_acc=np.zeros((Ks-1))
ConfustionMx=[];
for n in range(1,Ks):
    
    #Train Model and Predict  
    knn = KNeighborsClassifier(n_neighbors=n).fit(X_train,y_train)
    knn_y_pred = knn.predict(X_test)
    
    
    mean_acc[n-1]=np.mean(knn_y_pred==y_test);
    
    std_acc[n-1]=np.std(knn_y_pred==y_test)/np.sqrt(knn_y_pred.shape[0])
mean_acc

array([0.49712035, 0.48724299, 0.49834383, 0.49661305, 0.50628152,
       0.5069977 , 0.5069977 , 0.50687834, 0.50469995, 0.50643072,
       0.50455074, 0.49646385, 0.50064158, 0.49998508, 0.50004476,
       0.48813822, 0.49944794, 0.50884784, 0.50777357])

In [28]:
#Best k is 18
k = 18
knn = KNeighborsClassifier(n_neighbors = k).fit(X_train,y_train)

knn_y_pred = knn.predict(X_test)
knn_y_pred[0:5]

array([2, 2, 1, 2, 1])

In [29]:
print("KNN Jaccard index: %.2f" % jaccard_similarity_score(y_test, knn_y_pred))
print("KNN F1-score: %.2f" % f1_score(y_test, knn_y_pred, average='weighted') )

KNN Jaccard index: 0.51
KNN F1-score: 0.51


## Decision Tree

In [30]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(criterion="entropy", max_depth = 7)

dt.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=7,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [31]:
dt_y_pred = dt.predict(X_test)

In [32]:
print("DT Jaccard index: %.2f" % jaccard_similarity_score(y_test, dt_y_pred))
print("DT F1-score: %.2f" % f1_score(y_test, dt_y_pred, average='weighted') )

DT Jaccard index: 0.52
DT F1-score: 0.47


## Logisctic Regression

In [33]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C=6, solver='liblinear').fit(X_train,y_train)

In [34]:
LR_y_pred = LR.predict(X_test)

In [35]:
LR_y_prob = LR.predict_proba(X_test)

In [36]:
print("LR Jaccard index: %.2f" % jaccard_similarity_score(y_test, LR_y_pred))
print("LR F1-score: %.2f" % f1_score(y_test, LR_y_pred, average='weighted') )
print("LR LogLoss: %.2f" % log_loss(y_test, LR_y_prob))

LR Jaccard index: 0.52
LR F1-score: 0.51
LR LogLoss: 0.69


## Accuracy

In [37]:
from sklearn.metrics import accuracy_score
print("KNN Accuracy: ", accuracy_score(y_test, knn_y_pred))

KNN Accuracy:  0.5088478410074304


In [38]:
print("Decision Tree Accuracy: ", accuracy_score(y_test, dt_y_pred))

Decision Tree Accuracy:  0.5192921727194055


In [39]:
print("LR Accuracy: ", accuracy_score(y_test, LR_y_pred))

LR Accuracy:  0.5180686938617171


# Result 

| Algorithm | Jaccard | F1-score | LogLoss | Accuracy |
| --- | --- | --- | --- | --- |
| KNN | 0.51 | 0.51 | NA | 0.51 |
| DecisionTree | 0.52 | 0.47 | NA | 0.52 |
| LogisticRegression | 0.52 | 0.51 | 0.69 | 0.51 |